graph/validation_barrier.go

```diff
@@ -210,58 +285,176 @@ func (v *ValidationBarrier) WaitForDependants(job interface{}) error {
 	v.Lock()
 	defer v.Unlock()
 
 	switch msg := job.(type) {
-
-	// If we've just finished executing a ChannelAnnouncement, then we'll
-	// close out the signal, and remove the signal from the map of active
-	// ones. This will allow/deny any dependent jobs to continue execution.
 	case *lnwire.ChannelAnnouncement1:
-		finSignals, ok := v.chanAnnFinSignal[msg.ShortChannelID]
-		if ok {
-			if allow {
-				close(finSignals.allow)
-			} else {
-				close(finSignals.deny)
-			}
-			delete(v.chanAnnFinSignal, msg.ShortChannelID)
+		// Signal to the child jobs that parent validation has
+		// finished. We have to call removeParentJob for each annID
+		// that this ChannelAnnouncement can be associated with.
+		err := v.removeParentJob(msg.ShortChannelID, id)
+		if err != nil {
+			return err
+		}
+
+		err = v.removeParentJob(route.Vertex(msg.NodeID1), id)
+		if err != nil {
+			return err
+		}
+
+		err = v.removeParentJob(route.Vertex(msg.NodeID2), id)
+		if err != nil {
+			return err
 		}
 
-		delete(v.chanEdgeDependencies, msg.ShortChannelID)
+		return nil
 
-	// For all other job types, we'll delete the tracking entries from the
-	// map, as if we reach this point, then all dependants have already
-	// finished executing and we can proceed.
 	case *lnwire.NodeAnnouncement:
-		delete(v.nodeAnnDependencies, route.Vertex(msg.NodeID))
+		// Remove child job info.
+		v.removeChildJob(route.Vertex(msg.NodeID), id)
+		return nil
 	case *lnwire.ChannelUpdate1:
-		delete(v.chanEdgeDependencies, msg.ShortChannelID)
+		// Remove child job info.
+		v.removeChildJob(msg.ShortChannelID, id)
+		return nil
 
 	case *lnwire.AnnounceSignatures1:
-		return
+		// No dependency mappings are stored for AnnounceSignatures1,
+		// so do nothing.
+		return nil
+	}
+
+	return errors.New("invalid message - no job dependencies")
+}
+
+// removeParentJob removes parentJobID from the set of active parent jobs and
+// notifies to annID's child jobs that it has finished validating. This must be
+// called from SignalDependents.
+// NOTE: MUST be called with a mutex.
+func (v *ValidationBarrier) removeParentJob(annID any,
+	parentJobID JobID) error {
+
+	// Remove the parentJobID from activeParentJobIDs.
+	jobInfo, found := v.jobInfoMap[annID]
+	if !found {
+		// NOTE: Something seriously wrong has happened.
+		return fmt.Errorf("no job info found for identifier(%v)",
+			annID)
+	}
+
+	jobInfo.activeParentJobIDs.Remove(parentJobID)
+
+	lastJob := jobInfo.activeParentJobIDs.Empty()
+
+	// Notify all dependent jobs that a parent job has completed.
+	for child := range jobInfo.activeDependentJobs {
+		notifyChan, ok := v.childJobChans[child]
+		if !ok {
+			// Note: Something seriously wrong has happened.
+			return fmt.Errorf("no child job chan for child "+
+				"JobID(%v)", child)
+		}
+
+		// We don't want to block when sending out the signal.
+		select {
+		case notifyChan <- struct{}{}:
+		default:
+		}
```

Comment:
Are we ok with swallowing the signal instead? Seems like this could case jobs to never be run, particularly if `lastJob` is `true`

Reply:
I'm not sure what you mean here. 1. How would jobs not be run in the current scenario? 2. What does swallowing the signal look like here?